substrate/client/network/types/src/multihash.rs

```diff
@@ -0,0 +1,192 @@
+pub enum Error {
+	/// Invalid multihash size.
+	#[error("invalid multihash size '{0}'")]
+	InvalidSize(u64),
+	/// The multihash code is not supported.
+	#[error("unsupported multihash code '{0:x}'")]
+	UnsupportedCode(u64),
+	/// Catch-all for other errors emitted when converting `u64` code to enum or parsing multihash
+	/// from bytes. Never generated as of multihash-0.17.0.
+	#[error("other error: {0}")]
+	Other(Box<dyn std::error::Error + Send + Sync>),
+}
+
+impl From<LiteP2pError> for Error {
+	fn from(error: LiteP2pError) -> Self {
+		match error {
+			LiteP2pError::InvalidSize(s) => Self::InvalidSize(s),
+			LiteP2pError::UnsupportedCode(c) => Self::UnsupportedCode(c),
+			e => Self::Other(Box::new(e)),
+		}
+	}
+}
+
+impl From<Code> for LiteP2pCode {
+	fn from(code: Code) -> Self {
+		match code {
+			Code::Identity => LiteP2pCode::Identity,
+			Code::Sha2_256 => LiteP2pCode::Sha2_256,
+		}
+	}
+}
+
+impl TryFrom<LiteP2pCode> for Code {
+	type Error = Error;
+
+	fn try_from(code: LiteP2pCode) -> Result<Self, Self::Error> {
+		match code {
+			LiteP2pCode::Identity => Ok(Code::Identity),
+			LiteP2pCode::Sha2_256 => Ok(Code::Sha2_256),
+			_ => Err(Error::UnsupportedCode(code.into())),
+		}
+	}
+}
+
+impl TryFrom<u64> for Code {
+	type Error = Error;
+
+	fn try_from(code: u64) -> Result<Self, Self::Error> {
+		match LiteP2pCode::try_from(code) {
+			Ok(code) => code.try_into(),
+			Err(e) => Err(e.into()),
+		}
+	}
+}
+
+impl From<Code> for u64 {
+	fn from(code: Code) -> Self {
+		LiteP2pCode::from(code).into()
+	}
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
+pub struct Multihash {
+	multihash: LiteP2pMultihash,
+}
+
+impl Multihash {
+	/// Multihash code.
+	pub fn code(&self) -> u64 {
+		self.multihash.code()
+	}
+
+	/// Multihash digest.
+	pub fn digest(&self) -> &[u8] {
+		self.multihash.digest()
+	}
+
+	/// Wraps the digest in a multihash.
+	pub fn wrap(code: u64, input_digest: &[u8]) -> Result<Self, Error> {
+		LiteP2pMultihash::wrap(code, input_digest).map(Into::into).map_err(Into::into)
+	}
+
+	/// Parses a multihash from bytes.
+	///
+	/// You need to make sure the passed in bytes have the length of 64.
+	pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+		LiteP2pMultihash::from_bytes(bytes).map(Into::into).map_err(Into::into)
+	}
+
+	/// Returns the bytes of a multihash.
+	pub fn to_bytes(&self) -> Vec<u8> {
+		self.multihash.to_bytes()
+	}
+}
+
+/// Remove extra layer of nestedness by deferring to the wrapped value's [`Debug`].
+impl Debug for Multihash {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		Debug::fmt(&self.multihash, f)
+	}
+}
+
+impl From<LiteP2pMultihash> for Multihash {
+	fn from(multihash: LiteP2pMultihash) -> Self {
+		Multihash { multihash }
+	}
+}
+
+impl From<Multihash> for LiteP2pMultihash {
+	fn from(multihash: Multihash) -> Self {
+		multihash.multihash
+	}
+}
+
+// TODO: uncomment this after upgrading `multihash` crate to v0.19.1.
```

Comment:
Is there anything blocking us from updating to multihash 0.19.1? Maybe we could do that in a followup and raise an issue to not forget about it?

Reply:
Yes, it should be done at the same time as upgrading `libp2p`. I will mention it in the libp2p PR.